app/models/spree/virtual_gift_card.rb

```diff
@@ -33,16 +33,21 @@ def deactivated?
 
   def redeem(redeemer)
     return false if redeemed? || !redeemable?
-    create_store_credit!({
-      amount: amount,
-      currency: currency,
-      memo: memo,
-      user: redeemer,
-      created_by: redeemer,
-      action_originator: self,
-      category: store_credit_category,
-    })
-    self.update_attributes( redeemed_at: Time.now, redeemer: redeemer )
+
+    with_lock do
+      return false if redeemed? || !redeemable?
```

Comment:
A few suggestions

What if we extract `redeemed? || !redeemable?` into its own method, since it is being repeated
or
Is it good enough to return early from the lock and remove the guarding statement at the top of the method?

Reply:
Let's say we have:

1. thread 1 passes the first `redeemed? || !redeembable?`
2. thread 2 passes the first `redeemed? || !redeembable?`
3. thread 1 gets the lock
4. thread 2 waits
5. thread 1 completes the redemption (including setting `redeemed_at`)
6. thread 2 gets the lock (which also reloads the model)
7. if there's no check again in the lock, it would re-redeem the gift card

Now, technically only the check inside the `with_lock` is needed. But I also like to avoid locking if we don't have to. Which is why I have the lock outside as well. If there's a ton of connections hitting the same endpoint and some get to the method while others are still processing, they may just quickly fail instead of adding to the queue for the lock.

Note that the locks remain for the entirety of whatever transaction this is happening into. Which is why I usually try to avoid as much as possible explicit locking. That said, from the codebase, doesn't look like it's happening here.

That's basically my reasoning.

I am happy to remove the outside check if y'all prefer. Or extract to a separate method. Wasn't sure there was much of a point to that because it's a third party library and it wouldn't be something that we'd expose.

Reply:
I would also love to understand the motivation behind this. I know we have a couple of reports users able to redeem 2x - do you think the cause is because of multiple rapid clicks when redeeming from the front-end?